api/src/data_migration/load/load_oracle_data_task.py

```diff
@@ -0,0 +1,101 @@
+#
+# Load data from legacy (Oracle) tables to staging tables.
+#
+
+import logging
+
+import src.db.foreign
+import src.db.models.staging
+import src.task.task
+from src.adapters import db
+
+from . import sql
+
+logger = logging.getLogger(__package__)
+
+
+class LoadOracleDataTask(src.task.task.Task):
+    """Task to load data from legacy tables to staging tables."""
+
+    def __init__(self, db_session: db.Session, foreign_tables=None, staging_tables=None):
+        super().__init__(db_session)
+        if foreign_tables:
+            self.foreign_tables = foreign_tables
+        else:
+            self.foreign_tables = {t.name: t for t in src.db.foreign.metadata.tables.values()}
+        if staging_tables:
+            self.staging_tables = staging_tables
+        else:
+            self.staging_tables = {
+                t.name: t for t in src.db.models.staging.metadata.tables.values()
+            }
+
+    def run_task(self) -> None:
+        with self.db_session.begin():
+            self.load_data()
+
+    def load_data(self) -> None:
+        for table_name in self.foreign_tables:
+            self.do_staging_copy(table_name)
+
+    def do_staging_copy(self, table_name: str):
+        logger.info("process table", extra={"table": table_name})
+        foreign_table = self.foreign_tables[table_name]
+        staging_table = self.staging_tables[table_name]
+
+        self.log_row_count("row count before", foreign_table, staging_table)
+
+        insert_count = self.do_insert(foreign_table, staging_table)
+        update_count = self.do_update(foreign_table, staging_table)
+        logger.info(
+            "copy count",
+            extra={"table": table_name, "count.insert": insert_count, "count.update": update_count},
+        )
+
+        self.log_row_count("row count after", foreign_table, staging_table)
+
+    def do_insert(self, foreign_table, staging_table):
+        """Determine new rows by primary key, and copy them into the staging table."""
+
+        insert_from_select_sql, select_sql = sql.build_insert_select_sql(
+            foreign_table, staging_table
+        )
+
+        #print(insert_from_select_sql)
+
+        # COUNT has to be a separate query as INSERTs don't return a rowcount.
+        insert_count = self.db_session.query(select_sql.subquery()).count()
```

Comment:
Maybe as a follow-up if we want to improve performance, we could consider getting the count in the same query using `RETURNING` https://www.postgresql.org/docs/current/dml-returning.html

Did a little bit of testing locally - seems like its doable:
```sql
-- Sequence and defined type
CREATE SEQUENCE IF NOT EXISTS api.my_table_id_seq;

-- Table Definition
CREATE TABLE "api"."my_table" (
    "id" int4 NOT NULL DEFAULT nextval('api.my_table_id_seq'::regclass),
    "name" text,
    PRIMARY KEY ("id")

-- Actual query
with rows as (
INSERT INTO api.my_table (id, name)
		VALUES(5, 'bob'), (6, 'joe'), (7, 'steve')
	RETURNING
		id
	)
	select count(*) from rows
```
This gives `3` in the response.